eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Bbody.hs

```diff
@@ -308,3 +314,16 @@ instance
   where
   wrapFailed = LedgersFailure
   wrapEvent = ShelleyInAlonzoEvent . LedgersEvent
+
+totalRefScriptSizeInBlock ::
+  (EraTx era, BabbageEraTxBody era) => ProtVer -> StrictSeq (Tx era) -> UTxO era -> Int
+totalRefScriptSizeInBlock protVer txs (UTxO utxo)
+  | pvMajor protVer <= eraProtVerHigh @ConwayEra =
+      getSum $ foldMap' (Monoid.Sum . txNonDistinctRefScriptsSize (UTxO utxo)) txs
+  | otherwise =
+      snd $ F.foldl' accum (utxo, 0) txs
+  where
+    accum (!accUtxo, !accSum) tx =
+      let updatedUtxo = F.foldl' (\acc (k, v) -> Map.insert k v acc) accUtxo toAdd
+          toAdd = Map.toList $ unUTxO $ txouts (tx ^. bodyTxL)
```

Comment:
It is much more efficient to use `union` instead of manually inserting one element at a time. Most importantly, however you forgot to account for the `isValid` flag:
```suggestion
      let updatedUtxo = accUtxo `Map.union` toAdd
          toAdd = unUTxO $
            case tx ^. isValidTxL of
              IsValid True -> txouts (tx ^. bodyTxL)
              IsValid False -> collOuts (tx ^. bodyTxL)
```